src/Microsoft.AspNet.OData.Shared/Query/ODataQueryOptionsParserFactory.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using Microsoft.AspNet.OData.Adapters;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.AspNet.OData.Query
+{
+    /// <summary>
+    /// Factory for <see cref="IODataQueryOptionsParser"/> classes to handle parsing of OData query options in the request body.
+    /// </summary>
+    public static class ODataQueryOptionsParserFactory
+    {
+        /// <summary>
+        /// Creates a list of <see cref="IODataQueryOptionsParser"/>s to handle parsing of OData query options in the request body.
+        /// </summary>
+        /// <returns>A list of <see cref="IODataQueryOptionsParser"/>s to handle parsing of OData query options in the request body.</returns>
+        public static IList<IODataQueryOptionsParser> Create()
+        {
+            Type interfaceType = typeof(IODataQueryOptionsParser);
+            // Find all types that implement IODataQueryOptionsParser interface and have a parameterless constructor
+            var implementingTypes = TypeHelper.GetLoadedTypes(WebApiAssembliesResolver.Default).Where(type => type.IsClass &&
+                !type.IsAbstract &&
+                interfaceType.IsAssignableFrom(type) &&
+                type.GetConstructor(new Type[0]) != null);
+            
+            IList<IODataQueryOptionsParser> parsers = new List<IODataQueryOptionsParser>();
+
+            // Create an instance of each implementing type
+            foreach(var type in implementingTypes)
+            {
+                parsers.Add((IODataQueryOptionsParser)Activator.CreateInstance(type));
+            }
+
+            return parsers;
+        }
```

Comment:
Why is this method getting all parsers implementation in the assembly? If this is initializing the default, built-in parsers, then in my opinion the list should be created manually (like how the routing conventions does it). And if the user wants to add custom parser, then they should do so explicitly in the service provider, that way there's better control on which parsers are registered, and in what order. I believe the order matters, right?

Reply:
I considered that approach as well. See **Alternative Approaches and Enhancements** section in  the design document I shared internally.

Reply:
I agree that I'd rather have the parsers explicitly added rather than look across all loaded types to see if they implement IQueryOptionsParser.

---
In reply to: [437212714](https://github.com/OData/WebApi/pull/2191#discussion_r437212714) [](ancestors = 437212714)